Add DAGDependencyV2 and Converters

enavarro51 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Add new DAGDependencyV2 and Converters

### Details and comments

This PR is a replacement for #11310. In that PR, the new `DAGDependencyV2` was created and the associated template matching code was modified to use it. This PR only creates the dag and the 2 converters required. The following changes were made,
- There is a new DAGDependencyV2 which no longer uses DAGDepNode, but instead uses the existing DAGOpNode without modification. The makes for a cleaner and much reduced memory profile for the nodes.
- Additional methods from DAGCircuit, such as find_bit were added to the DAGDependencyV2 and other methods were modified to more closely match the 2 dags.
- The information stored previously in the node attributes would now be accessed in the template matching code through calls to the rustworkx functions once the template matching code is refactored.
- A terminology change was made in DAGDependencyV2. `direct_successors` is now `successors` and `successors` is now `descendants`. Similarly for `predecessors` and `ancestors`. This mirrors the usage in rustworkx.

Doing some local tests on an 8000 gate circuit and converting the circuit to a dag dependency, V2 used ~300MB of memory and V1 used ~2.5GB of memory.

The next phase will be to refactor the template matching code to use V2 and to optimize the matching.

*edit* (@jakelishman):

> [!WARNING]
>
> Just to highlight to anyone else reading this, again: all this is deliberately not public, and it's all so not public (undocumented, and has explicitly private names) that we're not even marking it with the "experimental" API. This is for near-term experimentation and there's a good chance it may even be replaced entirely with a pure-Rust implementation before being made public.

enavarro51 committed:
Fix typing-extensions

enavarro51 committed:
Merge branch 'main' of https://github.com/Qiskit/qiskit-terra into main

enavarro51 committed:
Merge branch 'main' of https://github.com/Qiskit/qiskit-terra into main

enavarro51 committed:
Merge branch 'main' of https://github.com/Qiskit/qiskit-terra into main

enavarro51 committed:
Add dagdependency_v2 and converters

enavarro51 committed:
Add V2 tests

enavarro51 committed:
Cleanup testing

enavarro51 committed:
Merge branch 'main' into dag_dep_v2

enavarro51 committed:
Move import

enavarro51 committed:
Merge branch 'dag_dep_v2' of github.com:enavarro51/qiskit-terra into dag_dep_v2

enavarro51 committed:
Add copy_empty_like and topological_op_nodes

enavarro51 committed:
Merge branch 'main' of https://github.com/Qiskit/qiskit-terra into main

enavarro51 committed:
Merge branch 'main' into dag_dep_v2

enavarro51 committed:
Lint and providers update

enavarro51 committed:
Change to apply_operation_back

enavarro51 committed:
Fix converter init

alexanderivrii commented on qiskit/dagcircuit/dagdependency_v2.py:
```diff
@@ -0,0 +1,740 @@
+            * all the qubits and clbits, including the registers.
+
+        Returns:
+            DAGCircuit: An empty copy of self.
```
It should return `DAGDependencyV2` :)

enavarro51 replied:
Correct.

enavarro51 replied:
Done in https://github.com/Qiskit/qiskit/pull/11705/commits/95affb7d8166a727673c07e51f7df886c34b9777.

alexanderivrii commented on qiskit/dagcircuit/dagdependency_v2.py:
```diff
@@ -0,0 +1,740 @@
+
+        # Maps of qarg/carg indices
+        self.qindices_map = {}
+        self.cindices_map = {}
```
What are these maps and why are they needed? It seems that these are not present either in `DAGCircuit`. Alternatively, maybe we need these in `DAGCircuit` as well?

enavarro51 replied:
These and all the code mentioned below that uses these in `apply_operation_back` is related to template matching. So if the goal is to ignore template matching for now and get the cleanest, closest to DAGCircuit version, these and associated code can all be removed.

enavarro51 replied:
Removed in https://github.com/Qiskit/qiskit/pull/11705/commits/95affb7d8166a727673c07e51f7df886c34b9777.

alexanderivrii commented on qiskit/dagcircuit/dagdependency_v2.py:
```diff
@@ -0,0 +1,740 @@
+        self.qindices_map = {}
+        self.cindices_map = {}
+
+    @property
+    def global_phase(self):
+        """Return the global phase of the circuit."""
+        return self._global_phase
+
```
In the future we should think on further unifying `DAGCircuit` and `DAGDependencyV2` functionality as there is a lot of shared code (related to global phases, calibrations, finding bits, and so on). Maybe we could have a class `DAGBase` or something like this that would implement the common functionality. Though this may be not completely straightforward as `DAGCircuit`s also have `DAGInNode`s and `DAGOutNode`s. What do you think?

enavarro51 replied:
This will probably require a more in-depth discussion, but on first look, there is very little that would be a problem with merging the 2 into a single class. Since `DAGDependencyV2` only uses op nodes, any checks in `DAGCircuit` for `DAGInNodes` would just be ignored. There are some methods like `add_wire` that would be different and any method like `compose` or `substitute_node` would have to deal with the possible commutation changes.

alexanderivrii commented on qiskit/dagcircuit/dagdependency_v2.py:
```diff
@@ -0,0 +1,740 @@
+        Function to copy a DAGDependencyV2 object.
+        Returns:
+            DAGDependencyV2: a copy of a DAGDependencyV2 object.
+        """
```
I am wondering if we need this function as `DAGCircuit` does not have it. 

If we do need the explicit `copy` function, can we copy the graph rather than having to reconstruct it from scratch (`apply_operation_back` checks commutativity between nodes and is generally expensive).

I also noticed that this does not copy the commutation library, which is an omission in the older `DAGDependency` implementation as well.

If `copy` is only required for template optimization, and the changes to template optimization will go to a follow-up PR, then maybe we can remove this function for now? (In addition, if I remember correctly, you have mentioned that we can avoid graph copying in template optimization).

enavarro51 replied:
Another one only used by template matching, so can remove for now.

enavarro51 replied:
Removed in https://github.com/Qiskit/qiskit/pull/11705/commits/95affb7d8166a727673c07e51f7df886c34b9777.

alexanderivrii commented on qiskit/dagcircuit/dagdependency_v2.py:
```diff
@@ -0,0 +1,740 @@
+        target_dag._global_phase = self._global_phase
+        target_dag.duration = self.duration
+        target_dag.unit = self.unit
+        target_dag.metadata = self.metadata
+        target_dag._key_cache = self._key_cache
```
We should probably also copy (the reference to) the commutation library. And what about `self.qindices_map` and `self.cindices_map`?

enavarro51 replied:
Don't need `qindices` or `cindices`. Can add `self.comm_checker`.

enavarro51 replied:
Removed and added in https://github.com/Qiskit/qiskit/pull/11705/commits/95affb7d8166a727673c07e51f7df886c34b9777.

alexanderivrii commented on qiskit/dagcircuit/dagdependency_v2.py:
```diff
@@ -0,0 +1,740 @@
+        params = tuple(params)
+        return (qubits, params) in self.calibrations[node.op.name]
+
+    def size(self):
+        """Returns the number of gates in the circuit"""
+        return len(self._multi_graph)
```
Just noting that `DAGCircuit` methods for `size`, `depth`, `count_ops` etc support additional arguments such as `recurse`. However, I think keep the `DAGDependency` methods simpler is completely fine for now.

alexanderivrii replied:
I think it would be useful to have the methods `num_qubits` and `num_clbits`.

enavarro51 replied:
Added along with `width` in https://github.com/Qiskit/qiskit/pull/11705/commits/95affb7d8166a727673c07e51f7df886c34b9777.

alexanderivrii commented on qiskit/dagcircuit/dagdependency_v2.py:
```diff
@@ -0,0 +1,740 @@
+    def predecessor_indices(self, node_id):
+        """Returns list of the indices of the predecessors of a node."""
+        # TODO: deprecate or remove this when template code no longer is based on node_id's
+        return sorted(self._multi_graph.predecessor_indices(node_id))
```
Let's not introduce these functions in this PR, since we may not ever need them.

enavarro51 replied:
Removed in https://github.com/Qiskit/qiskit/pull/11705/commits/95affb7d8166a727673c07e51f7df886c34b9777.

alexanderivrii commented on qiskit/dagcircuit/dagdependency_v2.py:
```diff
@@ -0,0 +1,740 @@
+            )
+        )
+
+    def ancestors(self, node):
+        """Returns set of the ancestors of a node as DAGOpNodes."""
+        return {self._multi_graph[x] for x in rx.ancestors(self._multi_graph, node._node_id)}
```
Just commenting on how amazing and clean it is to have the same interface with `DAGCircuit` and `rustworkx`! (And not the mess with `predecessors` meaning direct predecessors in one context and all the descendants in another).

enavarro51 replied:
Yes, I like that too.

alexanderivrii commented on qiskit/dagcircuit/dagdependency_v2.py:
```diff
@@ -0,0 +1,740 @@
+    def descendant_indices(self, node_id):
+        """Returns set of the indices of the descendants of a node."""
+        # TODO: deprecate or remove this when template code no longer is based on node_id's
+        return list(rx.descendants(self._multi_graph, node_id))
```
Same comment if we need these functions.

enavarro51 replied:
Removed in https://github.com/Qiskit/qiskit/pull/11705/commits/95affb7d8166a727673c07e51f7df886c34b9777.

alexanderivrii commented on qiskit/dagcircuit/dagdependency_v2.py:
```diff
@@ -0,0 +1,740 @@
+        named_nodes = []
+        for node in self._multi_graph.nodes():
+            if isinstance(node, DAGOpNode) and node.op.name in names:
+                named_nodes.append(node)
```
To be pedantic, we don't need `isinstance` check here.

enavarro51 replied:
Correct.

enavarro51 replied:
Done in https://github.com/Qiskit/qiskit/pull/11705/commits/95affb7d8166a727673c07e51f7df886c34b9777.

alexanderivrii commented on qiskit/dagcircuit/dagdependency_v2.py:
```diff
@@ -0,0 +1,740 @@
+                self.cindices_map[new_node] = []
+        else:
+            self.qindices_map[new_node] = []
+            self.cindices_map[new_node] = []
```
I am a bit confused about this block of code (which relates to my earlier confusion that I don't quite understand why we need `qindices_map` and `cindeices_map` in the first place). 

P.S. I understand that a variant of this code was present in the older `DAGDependency`, I am wondering if we can completely remove it for `V2`.

enavarro51 replied:
All will be removed if not needed for template matching.

enavarro51 replied:
Removed in https://github.com/Qiskit/qiskit/pull/11705/commits/95affb7d8166a727673c07e51f7df886c34b9777.

alexanderivrii commented on qiskit/dagcircuit/dagdependency_v2.py:
```diff
@@ -0,0 +1,740 @@
+        Returns:
+            node: corresponding to the label.
+        """
+        return self._multi_graph.get_node_data(node_id)
```
I am slightly confused: `DAGCircuit` does not have this function nor ever checks `_multi_graph.get_node_data`. Maybe we don't need these functions, given your changes to unifying the predecessor/successor interface? I.e. this looks like a point where the APIs for `DAGCircuit` and `DAGDependencyV2` slightly deviate, and I am wondering if we can avoid this.

enavarro51 replied:
This is needed in template matching so not important there, but also in `_update_edges` since that method is written using node indices instead of node object ids. I did a first try to convert this method to using object ids, but could not come up with an efficient way to do it. So for now I would leave this and address it later.

The use of node indices could also be a problem if nodes are inserted or removed and then iterated over.

enavarro51 replied:
Leaving this in until I rewrite `_update_edges` to not use indices. Also used in some of the tests, which will need to be redone later.

alexanderivrii replied:
Agreed. What about making `get_node` into an internal function (i.e. just renaming it to `_get_node`), so that we will able to freely change/remove it later?

alexanderivrii replied:
BTW, let's rewrite `_update_edges` in a follow-up PR, not now.

enavarro51 replied:
Agreed on `_get_node` and redo `_update_edges` later.

alexanderivrii commented on qiskit/dagcircuit/dagdependency_v2.py:
```diff
@@ -0,0 +1,740 @@
+
+    def remove_op_node(self, node):
+        """Remove an operation node n.
+
```
I am a little bit worried about this function in the context of DAG dependency. Though (after some thinking) I do believe that this method of removing nodes and reconnecting predecessors to successors is safe: we will not be able to incorrectly infer from dag dependency that two nodes commute when in fact they do not. It might not be compete: the dag dependency may say that two nodes do not commute when in fact they do. For example, suppose that we have a chain of nodes `A -> B -> C`. By definition, `A` and `B` do not commute, and `B` and `C` do not commute (but this says nothing about whether `A` and `C` commute). If we remove `B` and retain edges, we will end up with `A -> C`, which might be over-conservative if `A` and `C` do commute. Note that it's not enough to check the commutativity of `A` and `C` and remove the edge if they do commute, since we should also recursively check if `C` commutes with predecessors of `A`, and `A` commutes with successors of `C`. So at the minimum let's update the docstring of this function. Hmm, when would we ever want to remove nodes from a DAG?

enavarro51 replied:
The comment for `replace_block_with_op` addresses the question of commutativity of the output, which is relevant here as well. `remove_op_node` is used extensively in the transpiler passes.

alexanderivrii replied:
Agreed, `replace_block_with_op` has the same flavor: the resulting DAG-dependency may hide some commutativity relations. So at some point we may want to add a function to reconstruct `DAGDependency` from scratch.

alexanderivrii commented on qiskit/dagcircuit/dagdependency_v2.py:
```diff
@@ -0,0 +1,740 @@
+        Currently. this function is only used when creating a new DAGDependencyV2 from another
+        representation of a circuit, and hence there are no removed nodes (this is why
+        iterating over all nodes is fine).
+        """
```
This is just a warning: this function internally assumes that the newly added node has the largest node index and is added at the end of the graph. As we are extending the functionality of DAG dependency (and in particular adding methods to remove nodes), I am afraid that something here might break. Possibly it would be best to eventually rewrite this function to be more general.

enavarro51 replied:
As I mentioned above, this is one I took a first try at rewriting to use node object ids. I agree it should be redone to avoid iterating over the indices.

alexanderivrii reviewed (commented):
I am very excited about this new DAG-dependency interface because it's significantly closer to the `DAGCircuit` interface and in principle allows to unify the two even further. Ideally I would like the two interfaces to be almost fully identical (modulo the comment below) and I have left several inline comments to that extent. In particular, I am suggesting to remove or to adapt all methods that don't have a close `DAGCircuit` analogue (or add such methods to `DAGCircuit` as well). (This does not apply to internal methods prefixed with `_`).

But even right now this allows to write transpiler passes that can work with either circuit representation. 

Other than that, we want release notes and a few more tests exploring various `DAGDependencyV2` functionality (possibly many of these can be adapted from `DAGDependency` tests). 

The promised comment is that DAG-dependency is conceptually different from a DAG-circuit and some operations on DAG-dependency have to be implemented considerably differently. For instance, we don't yet have a DAG-dependency analogue of `substitute_node_with_dag` because adding new nodes to DAG dependency generally involves performing commutativity checks and recursively examining descendants and ancestors of the existing nodes.

enavarro51 committed:
Cleanup and remove indices methods

alexanderivrii commented on qiskit/dagcircuit/dagdependency_v2.py:
```diff
@@ -0,0 +1,688 @@
+        target_dag.metadata = self.metadata
+        target_dag._key_cache = self._key_cache
+        target_dag.comm_checker = CommutationChecker()
+
```
Can we do this instead: ``target_dag.comm_checker = self.comm_checker``?

enavarro51 replied:
Ok.

alexanderivrii commented on test/python/converters/test_circuit_to_dagdependency.py:
```diff
@@ -14,11 +14,11 @@
 from qiskit.converters.dagdependency_to_circuit import dagdependency_to_circuit
 from qiskit.converters.circuit_to_dagdependency import circuit_to_dagdependency
 from qiskit import QuantumRegister, ClassicalRegister, QuantumCircuit
-from test import QiskitTestCase  # pylint: disable=wrong-import-order
```
I can't say that I fully understand why PR #10998 replaced all lines of the form 
```
from test import QiskitTestCase
```
to
```
from test import QiskitTestCase  # pylint: disable=wrong-import-order
```
but it may be best to keep the change here.

enavarro51 replied:
Not sure why either, but can leave as is.

alexanderivrii commented on test/python/converters/test_circuit_to_dagdependency_v2.py:
```diff
@@ -0,0 +1,71 @@
+
+
+class TestCircuitToDagCanonical(QiskitTestCase):
+    """Test QuantumCircuit to DAGDependency."""
```
Should the name of the class be `TestCircuitToDagDependency` (not that it really matters).

enavarro51 replied:
Ok.

alexanderivrii reviewed (commented):
Thanks for the quick updates. I have added a few more minor inline suggestions, with the only other thing being the release notes :).

enavarro51 committed:
Change to _get_node and cleanup

enavarro51 committed:
Merge branch 'main' of https://github.com/Qiskit/qiskit-terra into main

enavarro51 committed:
Merge branch 'main' into dag_dep_v2

enavarro51 committed:
Add reno

enavarro51 commented:
Release note added and changes made.

enavarro51 committed:
Merge branch 'main' into dag_dep_v2

enavarro51 committed:
Fix node_id bug in dag_drawer

enavarro51 committed:
Merge branch 'dag_dep_v2' of github.com:enavarro51/qiskit-terra into dag_dep_v2

enavarro51 committed:
Merge branch 'main' of https://github.com/Qiskit/qiskit-terra into main

enavarro51 committed:
Merge branch 'main' into dag_dep_v2

enavarro51 committed:
Lint

enavarro51 committed:
Merge branch 'main' of https://github.com/Qiskit/qiskit-terra into main

enavarro51 committed:
Merge branch 'main' of https://github.com/Qiskit/qiskit-terra into main

enavarro51 committed:
Merge branch 'main' into dag_dep_v2

enavarro51 committed:
Convert to using leading _ for class and functions

enavarro51 committed:
Merge branch 'main' of https://github.com/Qiskit/qiskit-terra into main

enavarro51 committed:
Merge branch 'main' into dag_dep_v2

enavarro51 committed:
Cleanup test file import

enavarro51 committed:
Fix converters

enavarro51 committed:
Restart CI

enavarro51 committed:
Merge branch 'main' of https://github.com/Qiskit/qiskit-terra into main

enavarro51 committed:
Merge branch 'main' of https://github.com/Qiskit/qiskit-terra into main

sbrandhsn commented on qiskit/dagcircuit/dagdependency_v2.py:
```diff
@@ -0,0 +1,688 @@
+        if key is None:
+            key = _key
+
+        return iter(rx.lexicographical_topological_sort(self._multi_graph, key=_key))
```
```suggestion
        return iter(rx.lexicographical_topological_sort(self._multi_graph, key=key))
```
I just stumbled upon this typo while working on something else. :-)

jakelishman commented on releasenotes/notes/add-dagdependency-v2-7c591d708d6a8ac1.yaml:
Let's remove the release note, to keep it very much seeming internal.

enavarro51 replied:
Will remove the release note. As to other functions, I'll remove the `remove_op_node` and `remove_all_ops_named` functions. My question is, what does "all the functions that exist that we don't immediately have a use for" mean? There seems to be a difference of opinion on what that includes.

jakelishman reviewed (commented):
For others' reference: this `DAGDependencyV2` is exploratory work that we're doing with Ed (@enavarro51) about reworking the `DAGDependency` into a usable and much faster form for template matching, and potentially using it more throughout other parts of the transpiler pipeline.

We're looking to merge it into the `main` Qiskit as a fully private object, mostly to support longer-term working on the project with Ed as an external contributor while working with us.  There's a chance that the Python implementation never gets made fully public, but that instead we make a fully Rust version of this `DAGDependencyV2` (with Python wrappers) and use that in things like template matching and commutation analysis.  There are follow-up PRs to this that will do things like update the `_update_wires` function and the core data structure a little to include the requirement to track a `leaves` set when modifying the graph (in order to reduce the computational complexity of appends to the graph), but it's still in the largely exploratory phases at the moment.

Ed: I'm largely fine to merge this as-is, except for:
- let's remove the public release note
- we probably want to fix the typo Sebastian noticed

For the bits Sasha was talking about with `remove_op_node` (and really any removals): the best situation is probably just to remove those functions from this PR - if they're broken, there's no real reason to include them in even the exploratory merge.  I've no issue with removing _all_ the functions that exist that we don't immediately have a use for in the exploratory stuff - we can add useful things as/when we need them.  If anything, it's probably best to start with less than more, just so we don't forget to remove the cruft later.

enavarro51 committed:
Merge branch 'main' of https://github.com/Qiskit/qiskit-terra into main

enavarro51 committed:
Merge branch 'main' into dag_dep_v2

enavarro51 committed:
Remove reno and remove_ops functions and fix typo

enavarro51 committed:
Remove init entries for v2

jakelishman committed:
Remove meaningless methods

These `quantum_successors` etc have no meaning in `DAGDependency` (and
V2) since the edges don't carry the data that the `DAGCircuit` functions
check for; `DAGDependency` is in part a deliberate erasure of this data.

jakelishman reviewed (approved):
Thanks for this Ed.  I've removed the quantum/classical successor/predecessor methods because they were checking edge data that completely doesn't exist on `DAGDependency`.  I'm not super convinced that `replace_block_with_op` is completely right yet, but I think that might be the same as in `DAGDependency` and I don't want to erase it out from under Sasha.

Just to highlight to anyone else reading this, again: all this is deliberately not public, and it's all _so_ not public (undocumented, and has explicitly private names) that we're not even marking it with the "experimental" API.  This is for near-term experimentation and there's a good chance it may even be replaced entirely with a pure-Rust implementation before being made public.
